Add AutoscalingPolicy type definitions and autogenerated code

mattkelly opened the pull request:
### Description

 #### What does this pull request accomplish?

- Add `AutoscalingPolicy` type definitions and autogenerated code
- Update `MetricsBackend` to align with RFC updates
- Add codecov config to ignore autogenerated code

I'm mainly just breaking this stuff out to make an upcoming PR more easily reviewable.

 #### What issue(s) does this fix?
* None

 #### Additional Considerations

 ### Testing

It builds.

 ### Dependencies
 - None

ashleyschuett commented on pkg/apis/cerebral.containership.io/v1alpha1/types.go:
```diff
@@ -78,3 +79,50 @@ type AutoscalingGroupList struct {
+	MetricsBackend      string            `json:"metricsBackend"`
+	Metric              string            `json:"metric"`
+	MetricConfiguration map[string]string `json:"metricConfiguration"`
+	Policy              ScalingPolicy     `json:"policies"`
```
Should this be `Policies`? to match the json? Also do we want to match AutoscalingGroup and call this `ScalingPolicy` and then have the keys under be `Up` `Down`? Or should we update AutoscalingGroup to be `Strategy` with keys of `ScaleUp` and `ScaleDown`? I think I prefer the second, but either way I think it should be consistent? :/

mattkelly replied:
`policies` was a mistake. I kept flipping back and forth.

I'll change it to be `ScalingPolicy` with fields `ScaleUp` and `ScaleDown`. I'll update the ASG to match.

mattkelly commented:
Marked as `WIP` because I'll also add the CRD manifest with validation when I get a sec.

mattkelly commented on pkg/apis/cerebral.containership.io/v1alpha1/types.go:
```diff
@@ -78,3 +79,50 @@ type AutoscalingGroupList struct {
+	ComparisonOperator string  `json:"comparisonOperator"`
+	AdjustmentType     string  `json:"adjustmentType"`
+	AdjustmentValue    int     `json:"adjustmentValue"`
+	PollInterval       int     `json:"pollInterval"`
```
This and below belongs in parent 🤦‍♂️

mattkelly commented:
Updated:

- Added validation where it was missing and fixed validation where it was broken
- Added more examples to `example` dir for testing
- Fix consistency issues

mattkelly commented on deploy/crd/metrics-backend.yaml:
```diff
@@ -17,5 +17,14 @@ spec:
+          properties:
+            type:
+              type: string
+            configuration:
```
This is not required on purpose. It seems possible to one day have a metrics backend that requires no configuration.

mattkelly commented on pkg/apis/cerebral.containership.io/v1alpha1/types.go:
```diff
@@ -65,8 +66,8 @@ type AutoscalingGroupStatus struct {
 type ScalingStrategy struct {
-	Up   string `json:"up"`
-	Down string `json:"down"`
+	ScaleUp   string `json:"scapeUp"`
```
typo

mattkelly commented on deploy/crd/autoscaling-policy.yaml:
```diff
@@ -0,0 +1,65 @@
+    singular: autoscalingpolicy
+    plural: autoscalingpolicies
+    shortNames:
+    - asp
```
I personally liked the three letter names (despite the move from AutoScaling --> Autoscaling) ⚰️ but will change this to be consistent. We can revisit if we decide to later.

mattkelly replied:
Updated

normanjoyner replied:
Three letter abbreviation of autoscaling group is extremely common in the community. I would vote for 3 letter abbreviations of all of these if it's up for debate.

mattkelly replied:
Looks like that's 2-1 :) updated to make them both 3 letters.

ashleyschuett replied:
im fine with that, I only changed asg to ag since we changed `AutoScaling --> Autoscaling` as Matt mentioned. If a 3 letter abbreviation makes more sense lets change them back to `asp` and `asg` respectively. Sorry Matt :(
